### ISSUE TITLE
May fire event inconsistent behavior

### ISSUE THREAD

**Issue by lingceng** [aasm/aasm#294]
Hi, I have following code:

    class Person
      ...
      event :run, before: lambda { age = 10 }, guard: lambda { age > 5 }
    end

    person.age = 1

    person.may_run?
    # => false

    person.run
    # run sucessfully

Here `may_run?` is inconsitent with `run`.

I checked the source and found that:
- `may_run?` does not trigger  _before_ callback
- `run` triggers _before_ callback. So the guard get passed.

Is it a design purpose? Or a bug?

**Comment by anujaware** [comment]
@lingceng As per my understanding 'may_run' is just to check whether event is valid/ready to 'run' (or whatever event name is) or not. Callback will get triggered only when you actually fire the event.
Why do you want to fire callback without firing an event? I mean what is the scenario where you have such requirement?

**Comment by lingceng** [comment]
@anujaware Thank you for your reply. 
I do not want to fire callback when calling 'may_run?'. I want 'run' and 'may_run?' keep consistent. 
In a other word, I mean 'may_run?' return false 'run' should fail to run.

**Comment by anujaware** [comment]
Yes. But the example you have given will always pass as per the code written in before and guard. And I am not able to find the real time example where we do some change in before callback. If we do like this you can move it outside the event and as per the requirement first apply the required logic on object and then call may_run and run to keep it consistent.

I am trying to reproduce this but getting some errors. When I call run event on an object I am getting  AASM::InvalidTransition error. Can you please provide full code snippet from model. Is age is the field in the database or its just a variable? I want to check if I do some processing in before callback and if event fails does it rollback the changes or persist.

**Comment by lingceng** [comment]
@anujaware You're right. I can put the code outside `before` callback. 
Or do some tricks in the guard to keep the behavior consistent. eg. use age_was.
But in my opinion, guard should not be effected by any callbacks. So I asked whether it was a bug.

`age` is a field of Person. 
You should take care of initial state of a person to make `run` runnable. 
I'm afraid I can't provide the full code. It's just a fake snippet which simulates my real project logic.

But I can tell you the answer. 
No rollback happens in callbacks when you run with a non-bang method. 
Rollback happens in callbacks when you run with a bang method.  Be aware, though, that the objects will not have their instance data returned to their pre-transactional state. See more about transaction [here](http://apidock.com/rails/ActiveRecord/Transactions/ClassMethods)

**Comment by alto** [comment]
@lingceng That's how it's supposed to be. `may_run?` is used to check if the transition is actually "allowed" to run through, using guards only. `run` by itself then runs the transition and all other callbacks, which are meant to be like "do something else, too". Basically callbacks like `:before` should not have impact on the transition itself. They are meant to be triggers only.

@anujaware Thanks for your help!

**Comment by lingceng** [comment]
"Basically callbacks like :before should not have impact on the transition itself." is a way. But how about changing the callback stack as following?

      event           guards
      transition      guards
      event           before_all_events
      event           before

So the *before* callbacks never effect guards. So the *run* and *may_run* will always keep consistent.

**Comment by ngan** [comment]
I can see @lingchen's point being valid. Ehat he suggests would mirror active record. If you call `valid?` on an AR, it would run validations, in a sense, telling you if the record "may save." If you call `save` after that, it actually calls `valid?` again automatically. There's a lot of places in my code where I'm doing: `foo.run if may_run?`.

**Comment by pirj** [comment]
As per my tests it is consistent on branch master:

```
class Person
  include AASM

  attr_accessor :age

  def initialize age
    @age = age
  end

  aasm do
    state :sleeping, :initial => true
    state :running

    event :run, before: lambda { age = 10 }, guard: lambda { age > 5 } do
      transitions :from => :sleeping, :to => :running
      transitions :from => :running, :to => :finished
    end
  end
end

> p=Person.new 1
=> #<Person:0xa9fa114 @age=1>
> p.may_run?
=> false
> p.run
AASM::InvalidTransition: Event 'run' cannot transition from 'sleeping'. Failed callback(s): ["(pry)#17", "(pry)#17"].
from /home/pirj/source/oss/aasm/lib/aasm/aasm.rb:187:in `aasm_failed'

> p2=Person.new 7
=> #<Person:0xaa64988 @age=7>
> p2.may_run?
=> true
> p2.run
=> true
```

There's also a test to ensure that [callbacks does not run any state callback if the event guard fails](https://github.com/aasm/aasm/blob/master/spec/unit/callbacks_spec.rb#L130)

@lingceng can you please confirm that the issue has been fixed?

**Comment by pirj** [comment]
It seems to me that `event`'s `before` is kind of `before_validation` filter in ActiveRecord, but `may_run?` is not similar to AR's `valid?`, since it does not trigger `before`.
@lingceng how do you see it, should `before` be run before `may_run?` or should it not be run if `guard` fails?

**Comment by lingceng** [comment]
Both solution can make may_run? and run consistent. I choose it not be run if guard fails.
As @alto said, we expect the callback do not effect the guard.
